### PR TITLE
289 platform upgrade

### LINKED ISSUE #289: Add platform upgrade command
`esdc-platform-upgrade-diskboot`
```
Usage:   ./esdc-platform-upgrade-diskboot <new_platform_version> [--keep-smf-db]
Example: ./esdc-platform-upgrade-diskboot 20170624T192838Z
Args:
  --keep-smf-db         don't clear SMF database (useful if you don't want to loose
                        your manual changes to any SMF service properties)
```
`esdc-platform-upgrade-usbboot`
```
Usage:   ./esdc-platform-upgrade-usbboot <new_esdc_version>
Example: ./esdc-platform-upgrade-usbboot v2.6.7
```
The whole upgrade procedure can be managed like this:
```
cat /opt/erigones/bin/esdc-platform-upgrade
#!/bin/bash

if beadm list -H | grep -q dcos-; then
    /opt/erigones/bin/esdc-platform-upgrade-diskboot $PLATFORM_VER
elif /bin/bootparams | grep -q "^netboot=true"; then
    echo "This node is network booted. Nothing to do."
    exit 1
else
    /opt/erigones/bin/esdc-platform-upgrade-usbboot $ESDC_VER
fi
```

### PR DESCRIPTION
**Pull request by YanChii**
Closes  #289

### CODE REVIEW

File: bin/eslib/esdc-platform-upgrade-disk [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+			;;
+		*)
+			echo "Ignoring unknown argument '${1}'"
+			;;
+	esac
+	shift
+done
+
+if [[ "x${FORCE}" != "x1" ]] && [[ "${PLATFORM_VER}" == "$(uname -v | sed 's/^[a-z]*_//')" ]]; then
+	echo "The requested platform version is already running. Aborting upgrade."
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-disk]: Consider using `die()` from functions.sh.

File: bin/eslib/esdc-platform-upgrade-disk [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	lofi_remove "${USR_LOOPDEV}"
+	umount_path "${PLATFORM_MOUNT_DIR}"
+	lofi_remove "${PLATFORM_LOOPDEV}"
+	rm -rf "${UPG_DIR}"
+}
+upgrade_failed() {
+	printmsg "Upgrade failed!"
+	post_cleanup
+}
+trap upgrade_failed EXIT
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-disk]: Don't put a `trap` between function definitions. Please maintain a common order in a script (0. set 1. declarations/constants, 2. function definitions (internal first), ...)

The `trap` should go after the `### START ###`.

File: bin/eslib/esdc-platform-upgrade-disk [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+printmsg() {
+	echo "*** $* ***"
+}
+
+### START ###
+
+# pre-cleanup
+rm -rf "${UPG_DIR}"
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-disk]: This seems a bit nasty. Maybe it would be better to just fail if `/opt/upgrade` exists. If it exists it is not from this script (trap) and we should not delete user data without asking.

File: bin/eslib/esdc-platform-upgrade-disk [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+# correct the dcos-1
+# (issue https://github.com/erigones/esdc-factory/issues/94)
+if [[ "$(zfs get -Ho value canmount zones/ROOT/dcos-1)" != "noauto" ]]; then
+	zfs set canmount=noauto zones/ROOT/dcos-1
+fi
+
+# start download
+printmsg "Downloading the new platform"
+curl -k ${CURL_OPTS} "${PLATTFORM_DOWNLOAD_URL}" > "${PLATFORM_FILE}"
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-disk]: Use `-m` with some reasonable value (15?). Or `--connect-timeout`.
Use `-o` instead of `>`.
I also suggest adding the `-L` switch (support for future redirects).
Certificates on SmartOS work, so `-k` is not needed, right?

File: bin/eslib/esdc-platform-upgrade-disk [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+if [[ "${KEEP_SMF_DB}" != "1" ]]; then
+	# clear SMF database so it can be recreated at reboot
+	rm -f "${DCOS_MNTDIR}/etc/svc/repository.db"
+fi
+
+printmsg "Activating the new boot environment at next reboot"
+beadm activate "${NEW_BE}"
+trap post_cleanup EXIT
+printmsg "Upgrade completed successfuly"
+printmsg "Please reboot as soon as possible.."
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-disk]: Two dots at end of sentence.

File: bin/eslib/esdc-platform-upgrade-disk [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+cksum -x sha1 "${DCOS_MNTDIR}/platform/i86pc/amd64/boot_archive" | cut -d' ' -f1 > "${DCOS_MNTDIR}/platform/i86pc/amd64/boot_archive.hash"
+chown -R root:root "${DCOS_MNTDIR}/platform/i86pc/amd64"
+if [[ "${KEEP_SMF_DB}" != "1" ]]; then
+	# clear SMF database so it can be recreated at reboot
+	rm -f "${DCOS_MNTDIR}/etc/svc/repository.db"
+fi
+
+printmsg "Activating the new boot environment at next reboot"
+beadm activate "${NEW_BE}"
+trap post_cleanup EXIT
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-disk]: There must a better way.  Changing the trap complicates readability.
Please add only one trap and show the "Upgrade failed" message only if the upgrade fails.

File: bin/eslib/esdc-platform-upgrade-disk [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+printmsg "Creating a new boot environment"
+BEADM_NEW_NUMBER="$(expr 1 + $(beadm list -H | tail -1 | cut -d';' -f1 | cut -d- -f2))"
+if ! [[ $BEADM_NEW_NUMBER =~ ^[0-9]+$ ]]; then
+	# BEADM_NEW_NUMBER is not a correct number
+	echo "Cannot get number of current BE."
+	exit 5
+elif beadm list -H | cut -d';' -f1 | grep -q "dcos-${BEADM_NEW_NUMBER}"; then
+	# current BE already exists (beadm lists BE out of order?)
+	echo "Cannot get name of new BE."
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-disk]: Consider using `die()`

File: bin/eslib/esdc-platform-upgrade-disk [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+PLATFORM_LOOPDEV="$(lofiadm -a "${BOOT_ARCHIVE}")"
+mount -F ufs "${PLATFORM_LOOPDEV}" "${PLATFORM_MOUNT_DIR}"
+USR_LOOPDEV="$(lofiadm -a "${USR_ARCHIVE}")"
+mount -r -F ufs "${USR_LOOPDEV}" "${USR_ARCHIVE_MOUNT_DIR}"
+
+printmsg "Creating a new boot environment"
+BEADM_NEW_NUMBER="$(expr 1 + $(beadm list -H | tail -1 | cut -d';' -f1 | cut -d- -f2))"
+if ! [[ $BEADM_NEW_NUMBER =~ ^[0-9]+$ ]]; then
+	# BEADM_NEW_NUMBER is not a correct number
+	echo "Cannot get number of current BE."
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-disk]: Consider using `die()`

File: bin/eslib/esdc-platform-upgrade-disk [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+tar zxf "${PLATFORM_FILE}" -C "${PLATFORM_DIR}"
+
+printmsg "Accessing the new platform files"
+PLATFORM_LOOPDEV="$(lofiadm -a "${BOOT_ARCHIVE}")"
+mount -F ufs "${PLATFORM_LOOPDEV}" "${PLATFORM_MOUNT_DIR}"
+USR_LOOPDEV="$(lofiadm -a "${USR_ARCHIVE}")"
+mount -r -F ufs "${USR_LOOPDEV}" "${USR_ARCHIVE_MOUNT_DIR}"
+
+printmsg "Creating a new boot environment"
+BEADM_NEW_NUMBER="$(expr 1 + $(beadm list -H | tail -1 | cut -d';' -f1 | cut -d- -f2))"
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-disk]: This code block should be executed sooner. No need to download the new platform if you cannot calculate the BE name.

File: bin/eslib/esdc-platform-upgrade-disk [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+beadm mount "${NEW_BE}" "${DCOS_MNTDIR}"
+
+printmsg "Updating the new boot environment in the background"
+rsync -ac --delete --exclude=boot/loader.conf "${PLATFORM_MOUNT_DIR}/"{bin,boot,kernel,lib,platform,sbin,smartdc} "${DCOS_MNTDIR}"
+rsync -ac --delete "${USR_ARCHIVE_MOUNT_DIR}/"  "${DCOS_MNTDIR}/usr/"
+rsync -a --delete "${PLATFORM_DIR}/platform-${PLATFORM_VER}/i86pc/amd64" "${DCOS_MNTDIR}/platform/i86pc"
+# update boot_archive checksum
+cksum -x sha1 "${DCOS_MNTDIR}/platform/i86pc/amd64/boot_archive" | cut -d' ' -f1 > "${DCOS_MNTDIR}/platform/i86pc/amd64/boot_archive.hash"
+chown -R root:root "${DCOS_MNTDIR}/platform/i86pc/amd64"
+if [[ "${KEEP_SMF_DB}" != "1" ]]; then
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-disk]: Use integer comparison here.
Add a newline before the `if` statement to improve readability.

File: bin/eslib/esdc-platform-upgrade-disk [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+
+NEW_BE="dcos-${BEADM_NEW_NUMBER}"
+beadm create "${NEW_BE}"
+beadm mount "${NEW_BE}" "${DCOS_MNTDIR}"
+
+printmsg "Updating the new boot environment in the background"
+rsync -ac --delete --exclude=boot/loader.conf "${PLATFORM_MOUNT_DIR}/"{bin,boot,kernel,lib,platform,sbin,smartdc} "${DCOS_MNTDIR}"
+rsync -ac --delete "${USR_ARCHIVE_MOUNT_DIR}/"  "${DCOS_MNTDIR}/usr/"
+rsync -a --delete "${PLATFORM_DIR}/platform-${PLATFORM_VER}/i86pc/amd64" "${DCOS_MNTDIR}/platform/i86pc"
+# update boot_archive checksum
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-disk]: I think that this comment can be also printed to the user.

File: bin/eslib/esdc-platform-upgrade-disk [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+cksum -x sha1 "${DCOS_MNTDIR}/platform/i86pc/amd64/boot_archive" | cut -d' ' -f1 > "${DCOS_MNTDIR}/platform/i86pc/amd64/boot_archive.hash"
+chown -R root:root "${DCOS_MNTDIR}/platform/i86pc/amd64"
+if [[ "${KEEP_SMF_DB}" != "1" ]]; then
+	# clear SMF database so it can be recreated at reboot
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-disk]: This operation should be printed to the user.

File: bin/esdc-platform-upgrade [hunk 13]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash -e
+
+ESLIB="$(dirname $0)/eslib"
+
+# choose the right upgrade path for this node (disk, net, usb)
+if beadm list -H | grep -q dcos-; then
```
[the remark this answers is not in the document]
Comment (YanChii) [reply bin/esdc-platform-upgrade]: Refactored matching and creating new BE to be independent of the `dcos` prefix. Now the new BE name will be derrived from the current BE name regardless the prefix and number. 

If the trailing number exists, it will be incremented. 

If the trailing number does not exist, the `-1` will be added.

I've also added checks for BE name existence (in case we're upgrading from non-newest BE).

File: bin/esdc-platform-upgrade [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+    "${ESLIB}/esdc-platform-upgrade-disk" "$@"
+elif /bin/bootparams | grep -q "^netboot=true"; then
+    echo "This node is network booted. Nothing to do."
+    exit 0
+else
```
[the remark this answers is not in the document]
Comment (YanChii) [reply bin/esdc-platform-upgrade]: Additional checks added.

File: bin/eslib/esdc-platform-upgrade-disk (hunk 9 above)
Comment (YanChii) [reply bin/eslib/esdc-platform-upgrade-disk]: Now it fails in `esdc-platform-upgrade` if it cannot get the BE name. If the BE name can be obtained, the new BE name will be successful for sure.

File: bin/eslib/esdc-platform-upgrade-disk (hunk 4 above)
Comment (YanChii) [reply bin/eslib/esdc-platform-upgrade-disk]: CAs in Erigonos are not rolled out so AFAIK `-k` is needed.

**Comment by YanChii** [comment]
Multiple functions refactored into `functions.sh` and made globally reusable.

File: bin/eslib/esdc-platform-upgrade-disk (hunk 6 above)
Comment (YanChii) [reply bin/eslib/esdc-platform-upgrade-disk]: Introduced `FINISHED_SUCCESSFULLY` but personally I don't like it.

File: bin/eslib/esdc-platform-upgrade-disk (hunk 4 above)
Comment (dn0) [reply bin/eslib/esdc-platform-upgrade-disk]: Just use `/opt/local/bin/curl` instead of `/usr/bin/curl -k`.

File: bin/esdc-platform-upgrade [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+if dc_booted_from_hdd && [[ -n "$(_beadm_get_current_be_name)" ]]; then
+	# a boot env exists, this is disk install
+    "${ESLIB}/esdc-platform-upgrade-disk.sh" "$@"
+elif /bin/bootparams | grep -q "^netboot=true"; then
+    echo "This node is network booted. Nothing to do."
+    exit 0
+elif dc_booted_from_usb; then
+    "${ESLIB}/esdc-platform-upgrade-usb.sh" "$@"
+else
+	die 2 "Cannot determine platform boot type. Please contact support."
```
Comment (dn0) [inline bin/esdc-platform-upgrade]: Wrong indentation.
Please remove the message about contacting support.

File: bin/eslib/functions.sh [hunk 16]
```diff
@@ -52,6 +53,13 @@ BC=${BC:-"/usr/bin/bc"}
 QGA=${QGA:-"${ERIGONES_HOME}/bin/qga-client"}
 QGA_SNAPSHOT=${QGA_SNAPSHOT:-"${ERIGONES_HOME}/bin/qga-snapshot"}
 NODE=${NODE:-"/usr/node/bin/node"}
+BEADM=${BEADM:-"/usr/sbin/beadm"}
+RSYNC=${RSYNC:-"/usr/bin/rsync"}
+CURL=${CURL:-"/usr/bin/curl"}
```
Comment (dn0) [inline bin/eslib/functions.sh]: Are you sure you want use `/usr/bin/curl` by default?

File: bin/eslib/functions.sh [hunk 17]
```diff
@@ -163,6 +175,25 @@ round() {
 	printf "%.${places}f" "${number}"
 }
 
+umount_path() {
```
Comment (dn0) [inline bin/eslib/functions.sh]: There is a convention how all functions are written here. Please follow the convention:

Every positional argument is locally saved to a named variable.

File: bin/eslib/functions.sh [hunk 18]
```diff
@@ -163,6 +175,25 @@ round() {
...
+	if ${MOUNT} | grep -q "${1}"; then
```
Comment (dn0) [inline bin/eslib/functions.sh]: This won't work in some cases, e.g. you have nested mountpoints.

File: bin/esdc-platform-upgrade [hunk 19]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+set -e
+
+ESLIB="$(dirname $0)/eslib"
+
+. "${ESLIB}/functions.sh"
```
Comment (dn0) [inline bin/esdc-platform-upgrade]: The first thing to do is to determine ERIGONES_HOME. Then source it from there.
Please look at other scripts first. Consistency matters.

File: bin/eslib/functions.sh (hunk 18 above)
Comment (YanChii) [reply bin/eslib/functions.sh]: Yes.. in that case it will fail. This should be handled by user before calling of the function. Recursive umount should not happen unless explicitly specified (if it's really needed).

File: bin/eslib/esdc-platform-upgrade-disk.sh [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+	shift
+done
+
+# check platform version format
+if [[ ! "${PLATFORM_VER}" =~ ^[0-9]+T[0-9]+Z$ ]] && [[ ! "${PLATFORM_VER}" =~ ^v[0-9]+\.[0-9]+ ]]; then
+	die 2 "Unknown format of platform version: ${PLATFORM_VER}"
+fi
+
+# query platform version from esdc version if needed
+if [[ "${PLATFORM_VER}" =~ ^v[0-9]+\.[0-9]+ ]]; then
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-disk.sh]: We should also support version in other format - basically everything that begins with a v[0-9] should be considered a version string.

File: bin/eslib/esdc-platform-upgrade-disk.sh [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+${BEADM} create "${NEW_BE}"
+${BEADM} mount "${NEW_BE}" "${DCOS_MNTDIR}"
+
+printmsg "Updating the new boot environment in the background"
+${RSYNC} -ac --delete --exclude=boot/loader.conf "${PLATFORM_MOUNT_DIR}/"{bin,boot,kernel,lib,platform,sbin,smartdc} "${DCOS_MNTDIR}"
+${RSYNC} -ac --delete "${USR_ARCHIVE_MOUNT_DIR}/"  "${DCOS_MNTDIR}/usr/"
+${RSYNC} -a --delete "${PLATFORM_DIR}/platform-${PLATFORM_VER}/i86pc/amd64" "${DCOS_MNTDIR}/platform/i86pc"
+printmsg "Update boot_archive checksum"
+checksum "${DCOS_MNTDIR}/platform/i86pc/amd64/boot_archive" > "${DCOS_MNTDIR}/platform/i86pc/amd64/boot_archive.hash"
+chown -R root:root "${DCOS_MNTDIR}/platform/i86pc/amd64"
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-disk.sh]: Why is this needed?

File: bin/eslib/esdc-platform-upgrade-usb.sh [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+	echo "Usage:   $0 <new_dc_version> [-v] [-f]"
+	echo "Example: $0 v2.6.7"
+	echo "Args:"
+	echo "  -v                verbose download"
+	echo "  -f                force upgrade even when the requested version is already installed"
+
+	exit 1
+fi
+
+if ! [[ "${ESDC_VER}" =~ ^v[0-9]\.[0-9]\.[0-9]$ ]]; then
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-usb.sh]: This check is too strict. We should also support other version strings.

File: bin/eslib/esdc-platform-upgrade-usb.sh [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+CURR_USB_VER="$(cat "${USB_VERSION_FILE}" | sed -e 's/^esdc-.e-.n-//')"
+# shellcheck disable=SC2002
+WANTED_USB_IMG_VARIANT="$(cat "${USB_VERSION_FILE}" | sed -re 's/^(esdc-.e-.n-).*/\1/')"
+if [[ ${FORCE} -ne 1 ]] && [[ "${CURR_USB_VER}" == "${NEW_USB_VER}" ]]; then
+	die 0 "Requested ESDC version is already on the USB key. Nothing to do."
+fi
+
+umount_usb_key
+
+printmsg "Downloading new platform"
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-usb.sh]: new USB image

File: bin/eslib/esdc-platform-upgrade-usb.sh [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+ESDC_IMG="${WANTED_USB_IMG_VARIANT}${NEW_USB_VER}.img"
+ESDC_IMG_FULL="${UPG_DIR}/${ESDC_IMG}"
+ESDC_DOWNLOAD_URL="https://download.erigones.org/esdc/usb/stable/${ESDC_IMG}.gz"
+
+trap cleanup EXIT
+
+mkdir -p "${UPG_DIR}"
+# shellcheck disable=SC2086
+${CURL} ${CURL_OPTS} ${CURL_DEFAULT_OPTS} -o "${ESDC_IMG_FULL}.gz" "${ESDC_DOWNLOAD_URL}"
+printmsg "Unpacking new platform"
```
Comment (dn0) [inline bin/eslib/esdc-platform-upgrade-usb.sh]: platform -> USB image

File: bin/eslib/functions.sh [hunk 25]
```diff
@@ -821,3 +858,156 @@ _service_instance_delete() {
...
+
+_beadm_get_active_be_name() {
+	${BEADM} list -H 2>/dev/null | "${AWK}" -F';' '{if($3 == "R" || $3 == "NR") {print $1}}'
+}
+
+_beadm_check_be_exists() {
+	local BE="${1}"
+
+	${BEADM} list -H 2>/dev/null | grep -q -- "^${BE};"
+	return $?
```
Comment (dn0) [inline bin/eslib/functions.sh]: This return statement is redundant. Either

File: bin/eslib/functions.sh [hunk 26]
```diff
@@ -821,3 +858,156 @@ _service_instance_delete() {
...
+}
+
+_beadm_get_next_be_name() {
+	CURR_BE="$(_beadm_get_current_be_name)"
```
Comment (dn0) [inline bin/eslib/functions.sh]: missing local keyword + local variables should be lowercase

File: bin/eslib/functions.sh [hunk 27]
```diff
@@ -821,3 +858,156 @@ _service_instance_delete() {
...
+_beadm_get_current_be_name() {
+	${BEADM} list -H 2>/dev/null | "${AWK}" -F';' '{if($3 == "N" || $3 == "NR") {print $1}}'
+}
+
+_beadm_get_active_be_name() {
+	${BEADM} list -H 2>/dev/null | "${AWK}" -F';' '{if($3 == "R" || $3 == "NR") {print $1}}'
+}
+
+_beadm_check_be_exists() {
+	local BE="${1}"
```
Comment (dn0) [inline bin/eslib/functions.sh]: local variables should be lowercase

File: bin/eslib/functions.sh [hunk 28]
```diff
@@ -821,3 +858,156 @@ _service_instance_delete() {
...
+	echo "/mnt/$(svcprop -p 'joyentfs/usb_mountpoint' svc:/system/filesystem/smartdc:default)"
+}
+
+mount_usb_key() {
+	if [[ -n "$(_usbkey_get_mounted_path)" ]]; then
+		# already mounted
+		return 0
+	fi
+
+	local ALLDISKS=`/usr/bin/disklist -a`
```
Comment (dn0) [inline bin/eslib/functions.sh]: local variables should be lowercase.

File: bin/eslib/functions.sh [hunk 29]
```diff
@@ -821,3 +858,156 @@ _service_instance_delete() {
...
+	local USBMNT="$(_usbkey_get_mountpoint)"
+
+	mkdir -p "${USBMNT}"
+	for key in ${ALLDISKS}; do
+		if [[ "$(${FSTYP} /dev/dsk/${key}p1 2> /dev/null)" == 'pcfs' ]]; then
+			${MOUNT} -F pcfs -o foldcase,noatime /dev/dsk/${key}p1 ${USBMNT};
+			if [[ $? == "0" ]]; then
```
Comment (dn0) [inline bin/eslib/functions.sh]: use if command instead of $? comparision

File: bin/eslib/functions.sh [hunk 30]
```diff
@@ -821,3 +858,156 @@ _service_instance_delete() {
...
+		return 0
+	fi
+
+	local ALLDISKS=`/usr/bin/disklist -a`
+	local USBMNT="$(_usbkey_get_mountpoint)"
+
+	mkdir -p "${USBMNT}"
+	for key in ${ALLDISKS}; do
+		if [[ "$(${FSTYP} /dev/dsk/${key}p1 2> /dev/null)" == 'pcfs' ]]; then
+			${MOUNT} -F pcfs -o foldcase,noatime /dev/dsk/${key}p1 ${USBMNT};
```
Comment (dn0) [inline bin/eslib/functions.sh]: trailing ;

File: bin/eslib/esdc-platform-upgrade-disk.sh (hunk 21 above)
Comment (YanChii) [reply bin/eslib/esdc-platform-upgrade-disk.sh]: Because files in the platform are not root-owned. The owner is uid 1000.
```
root@n1:/var/tmp# ll platform-20170624T192838Z/i86pc/amd64/
total 546570
drwxr-xr-x   2 1000     1000           6 Jun 24 23:18 ./
drwxr-xr-x   4 1000     1000           4 Jun 24 23:12 ../
-rw-r--r--   1 1000     1000        266M Jun 24 23:17 boot_archive
-rw-r--r--   1 1000     1000        1.3K Jun 24 23:13 boot_archive.gitstatus
-rw-r--r--   1 1000     1000          41 Jun 24 23:18 boot_archive.hash
-rw-r--r--   1 1000     1000        1.2M Jun 24 23:13 boot_archive.manifest
```

Comment (dn0) [reply bin/eslib/esdc-platform-upgrade-disk.sh]: Then this is a bug in our PI build system.
